src/simulators/statevector/qubitvector.hpp

```diff
@@ -516,128 +684,120 @@ cvector_t QubitVector<statevector_t>::vector() const {
-    ss << num_states_ << "!=" << statevec.size() << ")";
-    throw std::runtime_error(ss.str());
-  }
-
-  const int_t end = num_states_;    // end for k loop
+template <class data_t>
+void QubitVector<data_t>::zero() {
+  const int_t end = size_;    // end for k loop
 
 #pragma omp parallel for if (num_qubits_ > omp_threshold_ && omp_threads_ > 1) num_threads(omp_threads_)
-  for (int_t k = 0; k < end; ++k)
-    statevector_[k] = statevec[k];
-}
-
-template <class statevector_t>
-void QubitVector<statevector_t>::initialize(const statevector_t &statevec, const size_t num_states) {
-  if (num_states_ != num_states) {
-    std::stringstream ss;
-    ss << "QubitVector<statevector_t>::initialize input vector is incorrect length (";
-    ss << num_states_ << "!=" << num_states << ")";
-    throw std::runtime_error(ss.str());
+  for (int_t k = 0; k < end; ++k) {
+    data_[k] = 0.0;
   }
-
-  const int_t end = num_states_;    // end for k loop
-
-#pragma omp parallel for if (num_qubits_ > omp_threshold_ && omp_threads_ > 1) num_threads(omp_threads_)
-  for (int_t k = 0; k < end; ++k)
-    statevector_[k] = statevec[k];
 }
 
-template <class statevector_t>
-void QubitVector<statevector_t>::set_num_qubits(size_t num_qubits) {
+template <class data_t>
+void QubitVector<data_t>::set_num_qubits(size_t num_qubits) {
   num_qubits_ = num_qubits;
-  num_states_ = 1ULL << num_qubits;
+  size_ = 1ULL << num_qubits;
 
   // Free any currently assigned memory
-  if (statevector_)
-    free(statevector_);
+  if (data_)
+    free(data_);
 
   if (checkpoint_) {
     free(checkpoint_);
     checkpoint_ = 0;
   }
 
   // Allocate memory for new vector
-  statevector_ = reinterpret_cast<complex_t*>(malloc(sizeof(complex_t) * num_states_));
+  data_ = reinterpret_cast<complex_t*>(malloc(sizeof(complex_t) * size_));
 }
 
-template <class statevector_t>
-double QubitVector<statevector_t>::norm() const {
-  double val = 0;
-  const int_t end = num_states_;    // end for k loop
-  #pragma omp parallel reduction(+:val) if (num_qubits_ > omp_threshold_ && omp_threads_ > 1) num_threads(omp_threads_)
-  {
-  #pragma omp for
-    for (int_t k = 0; k < end; k++)
-      val += std::real(statevector_[k] * std::conj(statevector_[k]));
-  } // end omp parallel
-  return val;
-}
 
-template <class statevector_t>
-void QubitVector<statevector_t>::checkpoint() {
+template <class data_t>
+void QubitVector<data_t>::checkpoint() {
   if (!checkpoint_)
-    checkpoint_ = reinterpret_cast<complex_t*>(malloc(sizeof(complex_t) * num_states_));
+    checkpoint_ = reinterpret_cast<complex_t*>(malloc(sizeof(complex_t) * size_));
 
-  const int_t end = num_states_;    // end for k loop
+  const int_t end = size_;    // end for k loop
 #pragma omp parallel for if (num_qubits_ > omp_threshold_ && omp_threads_ > 1) num_threads(omp_threads_)
   for (int_t k = 0; k < end; ++k)
-    checkpoint_[k] = statevector_[k];
+    checkpoint_[k] = data_[k];
```

Comment:
That would probably needs to be benchmarked... but I'd remove the OpenMP directive and put a `std::copy` instead of this loop.